services/construction_service.go

```diff
@@ -365,6 +369,54 @@ func (s *ConstructionAPIService) ConstructionMetadata(
 		return nil, wrapErr(ErrGeth, err)
 	}
 
+	// Build eth transaction for L1 fee calculation
+	var toAddress common.Address
+	var ethTx *ethTypes.Transaction
+	if to != "" {
+		toAddress = common.HexToAddress(to)
+	}
+	if isEIP1559 := gasTipCap != nil && gasFeeCap != nil; isEIP1559 {
+		ethTx = ethTypes.NewTx(&ethTypes.DynamicFeeTx{
+			Nonce:     nonce,
+			GasTipCap: gasTipCap,
+			GasFeeCap: gasFeeCap,
+			Gas:       gasLimit,
+			To:        &toAddress,
+			Value:     input.Value,
+			Data:      input.Data,
+		})
+	} else {
+		ethTx = ethTypes.NewTx(&ethTypes.LegacyTx{
+			Nonce:    nonce,
+			GasPrice: gasPrice,
+			Gas:      gasLimit,
+			To:       &toAddress,
+			Value:    input.Value,
+			Data:     input.Data,
+		})
+	}
+	ethTxBytes, err := ethTx.MarshalBinary()
+	if err != nil {
+		return nil, wrapErr(ErrL1DataFee, err)
+	}
+
+	// Get L1 data fee
+	var l1Fee *big.Int
+	if s.config.GethURL != "" {
+		client, err := ethclient.DialContext(ctx, s.config.GethURL)
+		if err != nil {
+			return nil, wrapErr(ErrL1DataFee, err)
+		}
+		gpoContract, err := bindings.NewGasPriceOracle(predeploys.GasPriceOracleAddr, client)
+		if err != nil {
+			return nil, wrapErr(ErrL1DataFee, err)
+		}
+		l1Fee, err = gpoContract.GetL1Fee(&bind.CallOpts{}, ethTxBytes)
```

Comment:
CallOpts contains a `Context` that should be set to `ctx`.

Reply:
done